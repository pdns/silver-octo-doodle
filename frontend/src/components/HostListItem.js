import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  withStyles,
  Avatar,
  Collapse,
  ListItem,
  ListItemIcon,
  ListItemText,
  ListItemSecondaryAction,
  Dialog,
  DialogTitle,
  DialogContent,
  Typography,
} from '@material-ui/core';
import {
  ExpandMore as ExpandIcon,
  ExpandLess as CollapseIcon,
} from '@material-ui/icons';
import DesktopIcon from 'mdi-material-ui/DesktopClassic';
import TargetList from './TargetList';
import HostMenu from './HostMenu';
import TargetInputForm from './TargetInputForm';
import { createTarget } from '../redux/actions';

const styles = {
  truncateOverflow: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  },
};

class HostListItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false,
      dialog: false,
    };
  }

  toggleExpand = () => {
    this.setState({ open: !this.state.open });
  }

  openDialog = () => {
    this.setState({ dialog: true });
  }

  closeDialog = () => {
    this.setState({ dialog: false });
  }

  handleCreate = (input) => {
    this.props.createTarget({
      variables: input,
      context: { hostId: this.props.host.id },
    });
    this.closeDialog();
  }

  render() {
    const { host, classes } = this.props;
    const { open } = this.state;
    const hasTargets = host.targetCount > 0;
    const icon = open ? <CollapseIcon /> : <ExpandIcon />;

    return (
      <React.Fragment>
        <ListItem button onClick={this.toggleExpand}>
          <Avatar>
            <DesktopIcon />
          </Avatar>
          <ListItemText className={classes.truncateOverflow}>
            <Typography className={classes.truncateOverflow}>
              {host.name}
            </Typography>
          </ListItemText>
          {hasTargets &&
            <ListItemIcon>
              {icon}
            </ListItemIcon>
          }
          <ListItemSecondaryAction>
            <HostMenu onCreate={this.openDialog} onView={() => { console.log('TODO'); }} />
          </ListItemSecondaryAction>
        </ListItem>
        {hasTargets &&
          <Collapse in={open} timeout="auto" unmountOnExit>
            <TargetList host={host} />
          </Collapse>
        }
        <Dialog open={this.state.dialog} fullWidth>
          <DialogTitle>
            Create new target
          </DialogTitle>
          <DialogContent>
            <TargetInputForm
              onCancel={this.closeDialog}
              onSubmit={this.handleCreate}
              hosts={[host]}
              defaultHost={host}
            />
          </DialogContent>
        </Dialog>
      </React.Fragment>
    );
  }
}

HostListItem.propTypes = {
  host: PropTypes.object.isRequired,
  createTarget: PropTypes.func,
  classes: PropTypes.object,
};

const mapDispatchToProps = {
  createTarget,
};

export default connect(
  null,
  mapDispatchToProps,
)(withStyles(styles)(HostListItem));
