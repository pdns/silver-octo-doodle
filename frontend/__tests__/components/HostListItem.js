/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HostListItem from '../../src/components/HostListItem';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react-redux', () => ({
  connect: () => x => x,
}));

jest.mock('@material-ui/core/Avatar', () => 'Avatar');
jest.mock('@material-ui/core/Dialog', () => 'Dialog');
jest.mock('@material-ui/core/Collapse', () => 'Collapse');
jest.mock('@material-ui/core/ListItem', () => 'ListItem');
jest.mock('@material-ui/core/DialogTitle', () => 'DialogTitle');
jest.mock('@material-ui/core/ListItemIcon', () => 'ListItemIcon');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('@material-ui/core/DialogContent', () => 'DialogContent');
jest.mock('@material-ui/core/ListItemSecondaryAction', () => 'ListItemSecondaryAction');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/icons/ExpandLess', () => 'CollapseIcon');
jest.mock('@material-ui/icons/ExpandMore', () => 'ExpandIcon');
jest.mock('mdi-material-ui/DesktopClassic', () => 'DesktopIcon');

jest.mock('../../src/components/TargetList', () => 'TargetList');
jest.mock('../../src/components/HostMenu', () => 'HostMenu');
jest.mock('../../src/components/TargetInputForm', () => 'TargetInputForm');

describe('HostListItem', () => {
  const createTargetMock = jest.fn();
  let host;
  let wrapper;

  beforeEach(() => {
    host = {
      id: 'my-host-id',
      name: 'my-host',
      targetCount: 9,
    };

    createTargetMock.mockReset();
    wrapper = shallow(<HostListItem host={host} createTarget={createTargetMock} />);
    wrapper = wrapper.dive();
  });

  it('should render correctly with no targets', () => {
    host.targetCount = 0;
    const tree = renderer
      .create(<HostListItem host={host} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with targets', () => {
    const tree = renderer
      .create(<HostListItem host={host} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should toggle Collapseable panel by clicking ListItem', () => {
    expect(wrapper.find('Collapse').prop('in')).toBe(false);
    wrapper.find('ListItem').simulate('click');
    expect(wrapper.find('Collapse').prop('in')).toBe(true);
    wrapper.find('ListItem').simulate('click');
    expect(wrapper.find('Collapse').prop('in')).toBe(false);
  });

  it('should open Dialog by calling function passed to HostMenu', () => {
    expect(wrapper.find('Dialog').prop('open')).toBe(false);
    wrapper.find('HostMenu').prop('onCreate')();
    expect(wrapper.find('Dialog').prop('open')).toBe(true);
  });

  it('should close Dialog by calling function passed to TargetInputForm.onCancel', () => {
    wrapper.find('HostMenu').prop('onCreate')();
    wrapper.find('TargetInputForm').prop('onCancel')();
    expect(wrapper.find('Dialog').prop('open')).toBe(false);
  });

  it('should call props.createTarget and close Dialog by calling function passed to TargetInputForm.onSubmit', () => {
    wrapper.find('HostMenu').prop('onCreate')();
    wrapper.find('TargetInputForm').prop('onSubmit')('yay');
    expect(createTargetMock).toHaveBeenCalledWith({ variables: 'yay', context: { hostId: 'my-host-id' } });
    expect(wrapper.find('Dialog').prop('open')).toBe(false);
  });
});
