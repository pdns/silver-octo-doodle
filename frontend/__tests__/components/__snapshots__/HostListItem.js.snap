// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostListItem should render correctly with no targets 1`] = `
Array [
  <ListItem
    button={true}
    onClick={[Function]}
  >
    <Avatar>
      <DesktopIcon />
    </Avatar>
    <ListItemText
      className="HostListItem-truncateOverflow-1"
    >
      <Typography
        className="HostListItem-truncateOverflow-1"
      >
        my-host
      </Typography>
    </ListItemText>
    <ListItemSecondaryAction>
      <HostMenu
        onCreate={[Function]}
        onView={[Function]}
      />
    </ListItemSecondaryAction>
  </ListItem>,
  <Dialog
    fullWidth={true}
    open={false}
  >
    <DialogTitle>
      Create new target
    </DialogTitle>
    <DialogContent>
      <TargetInputForm
        defaultHost={
          Object {
            "id": "my-host-id",
            "name": "my-host",
            "targetCount": 0,
          }
        }
        hosts={
          Array [
            Object {
              "id": "my-host-id",
              "name": "my-host",
              "targetCount": 0,
            },
          ]
        }
        onCancel={[Function]}
        onSubmit={[Function]}
      />
    </DialogContent>
  </Dialog>,
]
`;

exports[`HostListItem should render correctly with targets 1`] = `
Array [
  <ListItem
    button={true}
    onClick={[Function]}
  >
    <Avatar>
      <DesktopIcon />
    </Avatar>
    <ListItemText
      className="HostListItem-truncateOverflow-1"
    >
      <Typography
        className="HostListItem-truncateOverflow-1"
      >
        my-host
      </Typography>
    </ListItemText>
    <ListItemIcon>
      <ExpandIcon />
    </ListItemIcon>
    <ListItemSecondaryAction>
      <HostMenu
        onCreate={[Function]}
        onView={[Function]}
      />
    </ListItemSecondaryAction>
  </ListItem>,
  <Collapse
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <TargetList
      host={
        Object {
          "id": "my-host-id",
          "name": "my-host",
          "targetCount": 9,
        }
      }
    />
  </Collapse>,
  <Dialog
    fullWidth={true}
    open={false}
  >
    <DialogTitle>
      Create new target
    </DialogTitle>
    <DialogContent>
      <TargetInputForm
        defaultHost={
          Object {
            "id": "my-host-id",
            "name": "my-host",
            "targetCount": 9,
          }
        }
        hosts={
          Array [
            Object {
              "id": "my-host-id",
              "name": "my-host",
              "targetCount": 9,
            },
          ]
        }
        onCancel={[Function]}
        onSubmit={[Function]}
      />
    </DialogContent>
  </Dialog>,
]
`;
